File: CS50x/Week6_Python/ProblemSet6/sentimental-readability/readability.py
# import the needed libraries
from cs50 import get_string
import string


def main():
    # prompts user a text request;
    text = get_string("Text: ").strip().lower()

    # objects that store the result of the called functions, each of them receives
    # a string argument and returns an integer its results;
    letters = count_letters(text)
    words = count_words(text)
    sentences = count_sentences(text)

    # objects that receives the result of letters and sentences divided by words times 100
    l = (letters / words) * 100
    s = (sentences / words) * 100
    # object that stores the result of Coleman-Liau expression
    index = round((0.0588 * l) - (0.296 * s) - 15.8)

    # prompts user the result based on the value of the index;
    if index > 15:
        print("Grade 16+")
    elif index < 1:
        print("Before Grade 1")
    else:
        print(f"Grade {index}")


# function to count the valid letters of a sentence or word;
def count_letters(text):
    # objectic that stores letters without punctuation and spaces
    letters = text.translate(str.maketrans('', '', string.punctuation)).replace(" ", "")
    # object that store the numer of letters
    n = len(letters)
    return n

    """
    another way to do it:
    n = 0
        for l in text:
            if l.isalpha():
                n += 1
        return n
    """


# function to count the valid words of a sentence;
def count_words(text):
    # objectic that stores words without punctuation in a list;
    words = text.translate(str.maketrans('', '', string.punctuation)).split()
    # object that store the numer of letters founded in the list
    n = len(words)
    return n


# function to count the valid sentences of a text;
def count_sentences(text):
    # a list to store each sentence and 2 counters
    sentences = []
    c = -1
    i = 0

    # a loop to identify the determined punctuation and after that, will check if the previous
    # character is not a punctuation too, if it isn't, will store all the words before it in
    # an object, appending it in the sentences list
    for p in text:
        c += 1
        if p == '.' or p == '!' or p == '?':
            if text[c - 1] not in string.punctuation:
                pieces = text[i:c]
                sentences.append(pieces)
                i = c
    # object that store the numer of sentences founded in the list
    n = len(sentences)
    return n


main()

File: CS50x/Exercises/Week2_Arrays/ProblemSet2/readability/readability.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

//functions that receives a string argument and return an integer result;
int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int main(void)
{
    //prompts to user a text request;
    string text = get_string("Text: \n");
    //printf("%s\n", text);

    //a float variable that receives two integer functions, transform its values
    //in float results, calcutate its values and return a float result;
    float l = ((float)count_letters(text) / (float)count_words(text)) * 100;
    float s = ((float)count_sentences(text) / (float)count_words(text)) * 100;
    //an integer variable that calculates float values and round the result to
    //the nearest integer;
    int index = round((0.0588 * l) - (0.296 * s) - 15.8);

    //printf("%i letters\n", count_letters(text));
    //printf("%i words\n", count_words(text));
    //printf("%i sentences\n", count_sentences(text));

    //a condition that will print a text depending on the variable index result;
    if (index > 15)
    {
        printf("Grade 16+\n");
    }
    else if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }



}

//functions that receives a string argument and return an integer result;
int count_letters(string text)
{
    //a integer variable that store the number of text letters;
    int n = strlen(text);
    //a character array variable with the same position as the number of word letters;
    char filter[n];
    //an integer variable that will store the number of letters from capital A to Z;
    int letters = 0;

    //a loop that will scroll through all positions filter array allocating each value
    //of text array in it, also will change all letters to upper case;
    for (int i = 0; i < n; i++)
    {
        filter[i] = text[i];
        filter[i] = toupper(filter[i]);
    }

    //a loop that will count how many letters between A an Z there are and store it in letters;
    for (int j = 0; j < n; j++)
    {
        if (filter[j] > 64 && filter[j] < 91)
        {
            letters++;
        }
    }
    return letters;
}

//functions that receives a string argument and return an integer result;
int count_words(string text)
{
    //a integer variable that store the number of text letters;
    int n = strlen(text);
    //an integer variable that store the text number of words, assuming that the sentence
    //initiates with one word;
    int words = 1;

    //a loop that will scroll through all positions of text array searching the spaces
    //between words, each time it find one means that a new word starts past it;
    for (int i = 0; i < n; i++)
    {
        if (text[i] == 32)
        {
            words++;
        }
    }
    return words;
}

//functions that receives a string argument and return an integer result;
int count_sentences(string text)
{
    //a integer variable that store the number of text letters;
    int n = strlen(text);
    //an integer variable that store the text number of sentences;
    int sent = 0;

    //a loop that will scroll through all positions of text array searching the dots,
    //question and exclamation characters between words, each time it find one means that
    //theres a sentence before it;
    for (int i = 0; i < n; i++)
    {
        if (text[i] == 46 || text[i] == 33 || text[i] == 63)
        {
            sent++;
        }
    }
    return sent;
}